cub/device/dispatch/dispatch_batch_memcpy.cuh

```diff
@@ -0,0 +1,670 @@
+
+CUB_NAMESPACE_BEGIN
+
+namespace detail
+{
+
+/**
+ * Parameterizable tuning policy type for AgentBatchMemcpy
+ */
+template <uint32_t _BLOCK_THREADS, uint32_t _BYTES_PER_THREAD>
+struct AgentBatchMemcpyLargeBuffersPolicy
+{
+  enum
+  {
+    /// Threads per thread block
+    BLOCK_THREADS = _BLOCK_THREADS,
+    /// The number of bytes each thread copies
+    BYTES_PER_THREAD = _BYTES_PER_THREAD
+  };
+};
+
+/**
+ * Initialization kernel for tile status initialization (multi-block)
+ */
+template <typename BufferOffsetScanTileStateT, typename BlockOffsetScanTileStateT, typename TileOffsetT>
+__global__ void InitTileStateKernel(BufferOffsetScanTileStateT buffer_offset_scan_tile_state,
+                                    BlockOffsetScanTileStateT block_offset_scan_tile_state,
+                                    TileOffsetT num_tiles)
+{
+  // Initialize tile status
+  buffer_offset_scan_tile_state.InitializeStatus(num_tiles);
+  block_offset_scan_tile_state.InitializeStatus(num_tiles);
+}
+
+/**
+ * Kernel that copies buffers that need to be copied by at least one (and potentially many) thread blocks.
+ */
+template <typename ChainedPolicyT,
+          typename BufferOffsetT,
+          typename InputBufferIt,
+          typename OutputBufferIt,
+          typename BufferSizeIteratorT,
+          typename BufferTileOffsetItT,
+          typename TileT,
+          typename TileOffsetT>
+__launch_bounds__(int(ChainedPolicyT::ActivePolicy::AgentLargeBufferPolicyT::BLOCK_THREADS)) __global__
+  void MultiBlockBatchMemcpyKernel(InputBufferIt input_buffer_it,
+                                   OutputBufferIt output_buffer_it,
+                                   BufferSizeIteratorT buffer_sizes,
+                                   BufferTileOffsetItT buffer_tile_offsets,
+                                   TileT buffer_offset_tile,
+                                   TileOffsetT last_tile_offset)
+{
+  using StatusWord    = typename TileT::StatusWord;
+  using ActivePolicyT = typename ChainedPolicyT::ActivePolicy::AgentLargeBufferPolicyT;
+
+  constexpr uint32_t BLOCK_THREADS    = ActivePolicyT::BLOCK_THREADS;
+  constexpr uint32_t ITEMS_PER_THREAD = ActivePolicyT::BYTES_PER_THREAD;
+  constexpr uint32_t ITEMS_PER_BLOCK  = BLOCK_THREADS * ITEMS_PER_THREAD;
+
+  BufferOffsetT num_blev_buffers = buffer_offset_tile.LoadValid(last_tile_offset);
+
+  uint32_t tile_id = blockIdx.x;
+
+  // Make sure we've fetched read the total number of block-level buffers into num_blev_buffers
+  __syncthreads();
+
+  // No block-level buffers => we're done here
+  if (num_blev_buffers == 0)
+    return;
+
+  // While there's still tiles of bytes from block-level buffers to copied
+  do
+  {
+    __shared__ uint32_t buffer_id;
+
+    // Binary search the buffer that this tile belongs to
+    if (threadIdx.x == 0)
+      buffer_id = UpperBound(buffer_tile_offsets, num_blev_buffers, tile_id) - 1;
+
+    __syncthreads();
```

Comment:
For some unclear reasons we are using `CTA_SYNC` :smile:

Reply:
Replaced it with `CTA_SYNC` 👍